src/main/java/org/openlmis/integration/dhis2/web/dataset/DatasetController.java

```diff
@@ -0,0 +1,193 @@
+ */
+@Controller
+@RequestMapping(DatasetController.RESOURCE_PATH)
+@Transactional
+public class DatasetController extends BaseController {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DatasetController.class);
+
+  public static final String RESOURCE_PATH = ServerController.RESOURCE_PATH
+          + "/{serverId}" + "/datasets";
+
+  @Autowired
+  private DatasetRepository datasetRepository;
+
+  @Autowired
+  private ServerRepository serverRepository;
+
+  /**
+   * Retrieves the specified dataset.
+   */
+  @GetMapping(value = "/{id}")
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public DatasetDto getDataset(@PathVariable("id") UUID id) {
+    Dataset dataset = datasetRepository.findById(id)
+            .orElseThrow(() -> new NotFoundException(MessageKeys.ERROR_DATASET_NOT_FOUND));
+
+    return DatasetDto.newInstance(dataset);
+  }
+
+  /**
+   * Retrieves all datasets. Note that an empty collection rather than a 404 should be
+   * returned if no datasets exist.
+   */
+  @GetMapping
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public Page<DatasetDto> getAllDatasets(@PathVariable("serverId") UUID serverId,
+                                         Pageable pageable) {
+    Server server = serverRepository.findById(serverId)
+            .orElseThrow(() -> new NotFoundException(MessageKeys.ERROR_SERVER_NOT_FOUND));
+
+    List<Dataset> datasets = server.getDatasetList();
+    List<DatasetDto> datasetDtos = datasets
+            .stream()
+            .map(DatasetDto::newInstance)
+            .collect(Collectors.toList());
+
+    return Pagination.getPage(datasetDtos, pageable);
+  }
+
+  /**
+   * Allows the creation of a new dataset. If the id is specified, it will be ignored.
+   */
+  @PostMapping
+  @ResponseStatus(HttpStatus.CREATED)
+  @ResponseBody
+  public DatasetDto createDataset(@PathVariable("serverId") UUID serverId,
+                                  @RequestBody DatasetDto datasetDto) {
+    LOGGER.debug("Creating new dataset");
+    Dataset newDataset = Dataset.newInstance(datasetDto);
+
+    Server server = serverRepository.findById(serverId)
+            .orElseThrow(() -> new NotFoundException(MessageKeys.ERROR_SERVER_NOT_FOUND));
+    newDataset.setServer(server);
+
+    newDataset.setId(null);
+    newDataset = datasetRepository.saveAndFlush(newDataset);
+
+    return DatasetDto.newInstance(newDataset);
+  }
+
+  /**
+   * Updates the specified dataset.
+   */
+  @PutMapping(value = "/{id}")
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public DatasetDto updateDataset(@PathVariable("id") UUID id, @RequestBody DatasetDto datasetDto) {
+    if (null != datasetDto.getId() && !Objects.equals(datasetDto.getId(), id)) {
+      throw new ValidationMessageException(MessageKeys.ERROR_DATASET_ID_MISMATCH);
+    }
+
+    LOGGER.debug("Updating dataset");
+    Dataset dataset;
+    Optional<Dataset> datasetOptional = datasetRepository.findById(id);
```

Comment:
Or something like this, using Optional's methods:

```suggestion
    Dataset datasetToSave = datasetRepository.findById(id).map(dataset -> {
      dataset.updateFrom(datasetDto);
      return dataset;
    }).orElseGet(() -> Dataset.newInstance(datasetDto));
```

Reply:
Nice idea, changed in both dataset and server controller